# Firli Ilhami

## Apriori Algorithm Python
## Objective
Apriori is one of algorithms that has been used for Market Basket Analysis. Market basket analysis is process of discovering frequent item sets in large transactional database. In short, we want to know relationship beetween items that we have bought. In this case, we will use Burger Restaurant Dataset.

## Import Library

In [4]:
import numpy as np
import pandas as pd

How to install apyori on python:

!pip install apyori


In [5]:
from apyori import apriori

# Import Dataset
Our data is customers's transactions at Burger Restaurant in one day. <br> 
There are several features:
1. Timestamp
2. Type Burger
3. Code Burger
4. Amount Burger
5. Type Drink
6. Code Drink
7. Amount Drink
8. Gender
9. Payment



In [6]:
df=pd.read_csv('Burger Star.csv',delimiter=",")

In [7]:
df.head()

Timestamp           Type Burger  Code Burger  Amount Burger     Type Drink  \
0     10:01           Beef burger            1              6  Mineral water   
1     10:04         Cheese burger            3             12           Coke   
2     10:05  Double cheese burger            4             10  Mineral water   
3     10:06        Chicken burger            2              9          Coffe   
4     10:07         Cheese burger            3             11           Coke   

   Code Drink  Amount Drink  Gender   Payment  
0           1             3       1      DANA  
1           2            10       0      DANA  
2           1             5       0  LINK AJA  
3           3             9       0      DANA  
4           2             3       1      DANA

### Missing Value
There is no missing value in our dataset

In [9]:
df.isnull().sum()

Timestamp        0
Type Burger      0
Code Burger      0
Amount Burger    0
Type Drink       0
Code Drink       0
Amount Drink     0
Gender           0
Payment          0
dtype: int64

## Data Preparation
### We only use 'Type Burger' & 'Type Drink'
Type Burger :
1. Beef Burger
2. Cheese Burger
3. Double Cheese Burger
4. Cheese Burger

Type Drink :
1. Mineral Water
2. Coke
3. Coffe

We use only these features is because we just want to know which item has relationship with other items and we don't need amount item.

In [15]:
dataset=df[['Type Burger','Type Drink']]

In [16]:
dataset.head()

Type Burger     Type Drink
0           Beef burger  Mineral water
1         Cheese burger           Coke
2  Double cheese burger  Mineral water
3        Chicken burger          Coffe
4         Cheese burger           Coke

### Apriori algorithm  can only read  list, so we have to convert our dataset to list

In [17]:
transactions = []
for i in range(0, 312):
    transactions.append([str(dataset.values[i,j]) for j in range(0,2)])

In [18]:
transactions

[['Beef burger', 'Mineral water'],
 ['Cheese burger', 'Coke'],
 ['Double cheese burger', 'Mineral water'],
 ['Chicken burger', 'Coffe'],
 ['Cheese burger', 'Coke'],
 ['Double cheese burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Double cheese burger', 'Coffe'],
 ['Chicken burger', 'Coffe'],
 ['Cheese burger', 'Coke'],
 ['Beef burger', 'Mineral water'],
 ['Chicken burger', 'None'],
 ['Beef burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Beef burger', 'Coffe'],
 ['Beef burger', 'Coffe'],
 ['Chicken burger', 'Coffe'],
 ['Cheese burger', 'Coke'],
 ['Double cheese burger', 'Mineral water'],
 ['Double cheese burger', 'Coffe'],
 ['Double cheese burger', 'Mineral water'],
 ['Beef burger', 'None'],
 ['Chicken burger', 'None'],
 ['Cheese burger', 'Coke'],
 ['Double cheese burger', 'Mineral water'],
 ['Beef burger', 'Mineral water'],
 ['Double cheese burger',

## Modeling

###  Training Apriori on our dataset
We can set minimum support, minimum confidence, minimum lift and minimum length

In [19]:
from apyori import apriori
rules = apriori(transactions, min_support = 0.12, min_confidence = 0.0002, min_lift = 1.1, min_length = 1)

## Visualising The Results

In [20]:
results = list(rules)

In [21]:
for item in results:

    # first index of the inner list
    # Contains base item and add item
    pair = item[0] 
    items = [x for x in pair]
    print("Rule: " + items[0] + " -> " + items[1])

    #second index of the inner list
    print("Support: " + str(item[1]))

    #third index of the list located at 0th
    #of the third index of the inner list

    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("=====================================")

Rule: Beef burger -> Mineral water
Support: 0.20512820512820512
Confidence: 0.5517241379310345
Lift: 1.579247073710851
Rule: Coke -> Cheese burger
Support: 0.1794871794871795
Confidence: 1.0
Lift: 4.521739130434783
Rule: Double cheese burger -> Mineral water
Support: 0.14423076923076922
Confidence: 0.6521739130434783
Lift: 1.8667730355005985


## Result Explanation

### Rule : Beef burger -> Mineral water
* Support : 0.2051, it means probability customer buy Mineral water is 0.2051
* Confidence: 0.5517, it means probability customer buy Mineral water but they buy Beef burger first is 0.5517
* Lift : 1.579 , it means probability customer buy Mineral water but they buy Beef burger first is 1.579 times bigger than probability customer buy Mineral Water
* If lift is more than 1 , it means there are relationship between that items. So in this case, there are relationship between Beef burger and Mineral water
* Other rule has same meaning but with different object